src/Payments.Mvc/ClientApp/containers/CreateInvoiceContainer.tsx

```diff
@@ -0,0 +1,306 @@
+        this.setState({
+            [name]: value
+        });
+    }
+
+    private createNewItem = () => {
+        const items = this.state.items;
+        // needs new id logic
+        const id = items.byId.reduce((max, value) => Math.max(max, value), 0) + 1;
+        this.setState({
+            items: {
+                byHash: {
+                    ...items.byHash,
+                    [id]: {
+                        amount: 0,
+                        description: '',
+                        id,
+                        quantity: 0,
+                    },
+                },
+                byId: [...items.byId, id],
+            },
+        });
+    }
+
+    private removeItem = (id) => {
+        const items = this.state.items;
+        const newHash = {...items.byHash};
+        delete newHash[id];
+
+        const newItems = {
+            byHash: newHash,
+            byId: items.byId.filter(i => i !== id),
+        };
+
+        this.setState({
+            items: newItems,
+        });
+
+        // if this would set the list empty, add an empty one back
+        if (newItems.byId.length < 1) {
+            this.createNewItem();
+        }
+    }
+
+    private updateItem = (id, item) => {
+        const items = this.state.items;
+        const newHash = {...items.byHash};
+        newHash[id] = item;
+
+        this.setState({
+            items: {
+                byHash: newHash,
+                byId: items.byId,
+            },
+        });
+    }
+
+    private updateItemProperty = (id, name, value) => {
+        const item = this.state.items.byHash[id];
+        item[name] = value;
+        this.updateItem(id, item);
+    }
+
+    private calculateSubTotal = () => {
+        const items = this.state.items;
+        const sum = items.byId.reduce((prev, id) => {
+            const item = items.byHash[id];
+            return prev + (item.quantity * item.amount);
+        }, 0);
+
+        return sum;
+    }
+
+    private calculateTaxAmount = () => {
+        const { discount, taxRate } = this.state;
+        const sub = this.calculateSubTotal();
+        return (sub - discount) * taxRate;
+    }
+
+    private calculateTotal = () => {
+        const { discount } = this.state;
+        const sub = this.calculateSubTotal();
+        const tax = this.calculateTaxAmount();
+        return sub - discount + tax;
+    }
+
+    private onSubmit = async () => {
+        const { customers, discount, taxRate, items } = this.state;
+
+        const invoiceItems = items.byId.map(itemId => items.byHash[itemId]);
+
+        // create submit object
+        const invoice = {
+            customers,
+            discount,
+            items: invoiceItems,
+            tax: taxRate,
+        };
+
+        // create save url
+        const url = "/invoices/create";
+
+        // fetch 
+        const response = await fetch(url, {
+            body: JSON.stringify(invoice),
+            credentials: "include",
```

Comment:
@jpknoll same-origin instead of include?

Reply:
yeah, same-origin would probably be better.